src/accessor/mod.rs

```diff
@@ -0,0 +1,270 @@
+#![allow(missing_docs, clippy::missing_panics_doc, clippy::missing_errors_doc)]
+#[cfg(feature = "metadata")]
+use std::any::type_name;
+use std::{iter::FromIterator, pin::Pin};
+
+use futures_util::{
+	future::{err, ok},
+	Future, FutureExt,
+};
+
+#[cfg(feature = "metadata")]
+use crate::METADATA_KEY;
+use crate::{
+	atomics::Guard,
+	backend::Backend,
+	util::{is_metadata, InnerUnwrap},
+	Entry, IndexEntry, Key, Starchart,
+};
+
+type AccessorFuture<'a, O> = Pin<Box<dyn Future<Output = O> + Send + 'a>>;
+
+mod error;
+#[doc(hidden)]
+pub use self::error::{AccessorError, AccessorErrorType};
+
+/// A chart accessor used for passive access.
+#[cfg(feature = "passive")]
+#[must_use = "a ChartAccessor does nothing if not used."]
+pub struct ChartAccessor<'a, B> {
+	backend: &'a B,
+	guard: &'a Guard,
+}
+
+impl<'a, B> ChartAccessor<'a, B> {
+	pub(crate) const fn new(backend: &'a B, guard: &'a Guard) -> Self {
+		Self { backend, guard }
+	}
+}
+
+impl<'a, B: Backend> ChartAccessor<'a, B> {
+	pub async fn read_entry<S: Entry>(
```

Comment:
If we wanted to use an attribute to easily minify code sizes we could make this sync and use `futures_util::future::Either` to return an `impl Future` easily